ssl/test/bssl_shim.cc

```diff
@@ -173,9 +173,127 @@ class SocketCloser {
+static void MoveBIOs(SSL *dest, SSL *src) {
+  BIO *rbio = SSL_get_rbio(src);
+  BIO_up_ref(rbio);
+  SSL_set0_rbio(dest, rbio);
+
+  BIO *wbio = SSL_get_wbio(src);
+  BIO_up_ref(wbio);
+  SSL_set0_wbio(dest, wbio);
+
+  SSL_set0_rbio(src, nullptr);
+  SSL_set0_wbio(src, nullptr);
+}
+
+static bool TransferSSL(bssl::UniquePtr<SSL> *in, bssl::UniquePtr<SSL> *out) {
+  if (!in || !in->get()) {
+    return false;
+  }
+  SSL_CTX *in_ctx = SSL_get_SSL_CTX(in->get());
+  // Encode the SSL |in| into bytes.
+  // Decode the bytes into a new SSL.
+  bssl::UniquePtr<SSL> decoded_ssl;
+  if (!EncodeAndDecodeSSL(in->get(), in_ctx, &decoded_ssl)){
+    return false;
+  }
+  // Move the bio.
+  MoveBIOs(decoded_ssl.get(), in->get());
+  if (!SetTestConfig(decoded_ssl.get(), GetTestConfig(in->get()))) {
+    return false;
+  }
+  // Move the test state.
+  std::unique_ptr<TestState> state(GetTestState(in->get()));
+  if (!SetTestState(decoded_ssl.get(), std::move(state))) {
+    return false;
+  }
+  // Unset the test state of |in|.
+  std::unique_ptr<TestState> tmp1;
+  if (!SetTestState(in->get(), std::move(tmp1)) || !SetTestConfig(in->get(), nullptr)) {
+    return false;
+  }
+  // Free the SSL of |in|.
+  SSL_free(in->release());
+  // If |out| is not nullptr, |out| will hold the decoded SSL.
+  // Else, |in| will get reset to hold the decoded SSL.
+  if (out == nullptr) {
+    in->reset(decoded_ssl.release());
+  } else {
+    out->reset(decoded_ssl.release());
+  }
+  return true;
+}
+
+// Check if |in| can be encoded using |SSL_to_bytes|.
+static bool CanBeEncoded(const SSL *in) {
+  // |SSL_to_bytes| may generate new error code.
+  // |ERR_set_mark| and |ERR_pop_to_mark| are used to clean the error states.
+  ERR_set_mark();
+  // Encoding SSL to bytes.
+  size_t encoded_len;
+  bssl::UniquePtr<uint8_t> encoded;
+  uint8_t *encoded_raw;
+  bool ret = true;
+  if (SSL_to_bytes(in, &encoded_raw, &encoded_len)) {
+    encoded.reset(encoded_raw);
+  } else {
+    ret = false;
+    ERR_pop_to_mark();
+  }
+  return ret;
+}
+
+static void CheckSSLTransfer(const TestConfig *config, const SSL *ssl) {
+  if (config->check_ssl_transfer == 1 && CanBeEncoded(ssl)) {
+    // Below message is to inform runner.go that this test case can
+    // be converted to test SSL transfer.
+    // In the converted test, |CanBeEncoded| should be called again
+    // before |TransferSSL| because each test case may perform
+    // multiple connections. Not all connections can
+    // be encoded(identified by |CanBeEncoded| check).
+    fprintf(stderr, "Eligible for testing SSL transfer.\n");
+  }
+}
+
+static bool DoSSLTransfer(const TestConfig *config, bssl::UniquePtr<SSL> *in) {
+  if (config->ssl_transfer == 1 && CanBeEncoded(in->get())) {
+    // Below message is to inform runner.go that this test case 
+    // is going to test SSL transfer.
+    fprintf(stderr, "SSL transfer is going to be test.\n");
```

Comment:
Correct language according to comment in `runner.go`.

Reply:
Addressed.